zerver/lib/actions.py

```diff
@@ -3921,15 +3921,15 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
         event_last_message_id = get_last_message_id()
         all_subscription_logs = [
             RealmAuditLog(
-                realm=sub.user.realm,
+                realm=sub_info.user.realm,
                 acting_user=acting_user,
                 modified_user=sub_info.user,
                 modified_stream=sub_info.stream,
                 event_last_message_id=event_last_message_id,
                 event_type=RealmAuditLog.SUBSCRIPTION_DEACTIVATED,
                 event_time=event_time,
             )
-            for sub in subs_to_deactivate
+            for sub_info in subs_to_deactivate
         ]
```

Comment:
Good find!  @mateuszmandera @alexmv FYI.  Note that because the loop variable was wrong, the actual lines that were incorrect in past events are `modified_user` and `modified_stream`

This likely explain the situation we were trying to debug with some users having `SUBSCRIPTION_DEACTIVATED` entries that didn't make sense, which is unfortunate, because we use that data in several places.

It's not clear to me to what extent we have the data to uncorrupt this aspect of the database.  It seems possible that we could identify records that may be corrupt at least, because:
(1) The bug only affects cases where subscriptions were bulk-removed
(2) We'll have a batch of events all with the same `event_time`, `acting_user`, and `event_last_message_id`.
(3) I'm not sure any of our official clients actually use the API endpoint to bulk-remove users from a stream in a single request, so it's possible that this is rare enough we could investigate and manually fix, at least for Zulip Cloud.

Reply:
While investigating this, I noticed that we have various `RealmAuditLog` entries with `event_last_message_id` being some implausibly low message ID for Zulip Cloud, like 3200, for realms that we definitely imported from a self-hosted Zulip server.  I think the bug is this:
```
diff --git a/zerver/lib/import_realm.py b/zerver/lib/import_realm.py
index 9e2bbc1118..04ba7c6941 100644
--- a/zerver/lib/import_realm.py
+++ b/zerver/lib/import_realm.py
@@ -1035,6 +1035,9 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
             data, "zerver_realmauditlog", "acting_user", related_table="user_profile"
         )
         re_map_foreign_keys(data, "zerver_realmauditlog", "modified_stream", related_table="stream")
+        re_map_foreign_keys(
+            data, "zerver_realmauditlog", "event_last_message_id", related_table="message"
+        )
         update_model_ids(RealmAuditLog, data, related_table="realmauditlog")
         bulk_import_model(data, RealmAuditLog)
     else:
```

(totally untested) -- i.e. that we're not remapping this field's ID namespace.  

@hackerkid does that seem right to you?  If so, would you be up for doing a separate PR to prevent that from happening in the future?

Reply:
Opened https://github.com/zulip/zulip/issues/18621 to track the cleanup work related to this bug.